src/inmanta/execute/dataflow/root_cause.py

```diff
@@ -0,0 +1,134 @@
+"""
+    Copyright 2020 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+from itertools import chain
+from typing import Iterable, Iterator, List, Optional, Set
+
+from inmanta.execute.dataflow import AssignableNodeReference, AttributeNode, AttributeNodeReference, InstanceNode
+
+
+class RootCauseAnalyzer:
+    """
+        Analyzes the root causes among a collection of attribute nodes. The main entrypoint for this class is the
+        root_causes method.
+    """
+
+    def __init__(self, nodes: Iterable[AttributeNode]) -> None:
+        self.nodes: Set[AttributeNode] = set(nodes)
+
+    def root_causes(self) -> Set[AttributeNode]:
+        """
+            Returns the root causes from this instances' set of attribute nodes. An attribute node c
+            is defined as the cause for an other attribute node n iff c being unset leads to n
+            being unset. For a formal definition, see is_cause.
+        """
+
+        def caused_by(node: AttributeNode, others: Set[AttributeNode]) -> bool:
+            return any(self.is_cause(other, node) for other in others)
+
+        return set(node for node in self.nodes if not caused_by(node, self.nodes.difference({node})))
+
+    def is_cause(self, cause: AttributeNode, node: AttributeNode) -> bool:
+        """
+            Returns True iff cause is a cause for node.
```

Comment:
Could you not shortcut a LOT of work by taking into account that any node that has a value can never be the cause of an unset exception?

Reply:
Very good point, didn't think of that.